core/src/main/java/org/infinispan/expiration/ExpirationManager.java

```diff
@@ -35,11 +35,32 @@
    boolean isEnabled();
 
    /**
-    * This should be invoked passing in an entry that is now expired.  Note this method assumes to not have the
-    * lock for this key and will attempt to acquire it.
+    * This should be invoked passing in an entry that is now expired.  This method assumes the lock is not acquire
+    * and may attempt to lock this key to preserve atomicity.
     * @param entry entry that is now expired
     */
    void handleInMemoryExpiration(InternalCacheEntry<K, V> entry);
 
+   /**
+    * This is to be invoked when a store entry expires.  This method assumes the lock is not acquire and may attempt
+    * to lock this key to preserve atomicity.
+    * <p>
+    * Note this method doesn't currently take a {@link InternalCacheEntry} and this is due to a limitation in the
+    * cache store API.  This may cause some values to be removed if they were updated at the same time.
```

Comment:
The problem is the CacheStores are provided a PurgeListener which only takes a key.  I had thought about removing usage of the PurgeListener and instead inject the ExpirationManager into the TaskContext, but that only applies to process method as well.  I couldn't figure out anything nice, unless you have any suggestions?